File: server/atlas-evt-receiver.js
'use strict';
/* Atlas Analytics receiver (P0-004) — zero-dependency Node HTTP endpoint.
   POST /api/atlas/evt  { aid, t, name, payload }

   Contract (mirrors web/analytics.js):
   - CORS: only https://www.f8w.com and https://atlas.f8w.com, no credentials
   - body limit 1KB, strict event-name whitelist, field whitelist + length caps
   - appends one JSON line per day file: data/atlas-events-YYYY-MM-DD.jsonl
   - never stores anything beyond the filtered envelope (Prompt text cannot
     pass the filters even if a buggy client tried to send it)

   Run:  node server/atlas-evt-receiver.js            (default port 8901)
         PORT=8901 node server/atlas-evt-receiver.js
   Production: put behind nginx (TLS termination) — see server/README.md */

const http = require('http');
const fs = require('fs');
const path = require('path');

const PORT = Number(process.env.PORT) || 8901;
const ALLOWED_ORIGINS = ['https://www.f8w.com', 'https://atlas.f8w.com'];
const MAX_BODY = 1024;
const MAX_STR = 60;
const MAX_ARRAY = 30;

const EVENT_FIELDS = {
  page_view: ['path', 'ref', 'lang'],
  check_started: ['mode', 'model', 'scene', 'promptLength'],
  check_completed: ['foundAtoms', 'controlLevel', 'coverage', 'invalidCount'],
  evidence_viewed: ['atomId', 'kind'],
  recommendation_applied: ['atomId', 'slot', 'source'],
  export_clicked: ['kind', 'atomsCount'],
  share_created: ['kind', 'score'],
  feedback_vote: ['atomId', 'vote']
};
const EXPORT_KINDS = ['prompt', 'optimized_prompt', 'recipe', 'storyboard'];
const SHARE_KINDS = ['check_report', 'recipe_card', 'atom_card', 'storyboard_card'];

const AID_RE = /^[A-Za-z0-9-]{8,64}$/;

function cors(req, res) {
  const origin = req.headers.origin;
  if (origin && ALLOWED_ORIGINS.indexOf(origin) !== -1) {
    res.setHeader('Access-Control-Allow-Origin', origin);
    res.setHeader('Vary', 'Origin');
  }
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
  // deliberately NO Access-Control-Allow-Credentials
}

function filterEvent(name, payload) {
  const fields = EVENT_FIELDS[name];
  if (!fields) return null;
  const out = {};
  for (const k of fields) {
    const v = payload ? payload[k] : undefined;
    if (v === undefined || v === null) continue;
    if (typeof v === 'string') {
      if (!v || v.length > MAX_STR) return null;
      out[k] = v;
    } else if (typeof v === 'number' && isFinite(v)) {
      out[k] = v;
    } else if (Array.isArray(v)) {
      if (!v.length || v.length > MAX_ARRAY) return null;
      if (v.some(x => typeof x !== 'string' || !x || x.length > MAX_STR)) return null;
      out[k] = v;
    }
  }
  if (name === 'export_clicked' && !EXPORT_KINDS.includes(out.kind)) return null;
  if (name === 'share_created' && !SHARE_KINDS.includes(out.kind)) return null;
  return out;
}

function appendEvent(line) {
  const dir = path.join(__dirname, 'data');
  fs.mkdirSync(dir, { recursive: true });
  const day = new Date().toISOString().slice(0, 10);
  fs.appendFileSync(path.join(dir, 'atlas-events-' + day + '.jsonl'), line + '\n');
}

const server = http.createServer(function (req, res) {
  cors(req, res);
  if (req.method === 'OPTIONS') { res.writeHead(204); res.end(); return; }
  if (req.method !== 'POST' || req.url.split('?')[0] !== '/api/atlas/evt') {
    res.writeHead(404, { 'Content-Type': 'text/plain' });
    res.end('not found');
    return;
  }
  let body = '';
  let tooLarge = false;
  req.on('data', function (chunk) {
    body += chunk;
    if (body.length > MAX_BODY) { tooLarge = true; req.destroy(); }
  });
  req.on('end', function () {
    if (tooLarge) return; // connection already torn down
    let json = null;
    try { json = JSON.parse(body); } catch (e) { /* fallthrough */ }
    if (!json || typeof json !== 'object'
      || typeof json.aid !== 'string' || !AID_RE.test(json.aid)
      || typeof json.name !== 'string'
      || !json.payload || typeof json.payload !== 'object') {
      res.writeHead(400, { 'Content-Type': 'text/plain' });
      res.end('bad request');
      return;
    }
    const filtered = filterEvent(json.name, json.payload);
    if (!filtered) {
      res.writeHead(422, { 'Content-Type': 'text/plain' });
      res.end('invalid event');
      return;
    }
    const record = {
      aid: json.aid,
      t: typeof json.t === 'number' && isFinite(json.t) ? json.t : Date.now(),
      name: json.name,
      payload: filtered,
      recvAt: new Date().toISOString()
    };
    try {
      appendEvent(JSON.stringify(record));
      res.writeHead(204);
      res.end();
    } catch (e) {
      res.writeHead(500, { 'Content-Type': 'text/plain' });
      res.end('storage error');
    }
  });
});

server.listen(PORT, function () {
  console.log('atlas-evt-receiver listening on :' + PORT + ' (POST /api/atlas/evt)');
});

File: server/atlas-evt-receiver.py
#!/usr/bin/env python3
"""Atlas Analytics receiver (Python port, deployed on www.f8w.com).

Contract identical to server/atlas-evt-receiver.js:
- POST /api/atlas/evt  {aid, t, name, payload}
- CORS: only https://www.f8w.com / https://atlas.f8w.com, NO credentials
- body limit 1KB, strict event/field whitelists, strings > 60 chars or
  arrays > 30 items invalidate the whole event (Prompt text can never pass)
- appends one JSON line per day: data/atlas-events-YYYY-MM-DD.jsonl

Run:  /usr/bin/python3 /opt/atlas/atlas-evt-receiver.py   (PORT env, default 8901)
"""
import json
import os
import re
import threading
from datetime import datetime, timezone
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import urlsplit

PORT = int(os.environ.get('PORT', '8901'))
DATA_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'data')
ALLOWED_ORIGINS = ('https://www.f8w.com', 'https://atlas.f8w.com')
MAX_BODY = 1024
MAX_STR = 60
MAX_ARRAY = 30

EVENT_FIELDS = {
    'page_view': ('path', 'ref', 'lang'),
    'check_started': ('mode', 'model', 'scene', 'promptLength'),
    'check_completed': ('foundAtoms', 'controlLevel', 'coverage', 'invalidCount'),
    'evidence_viewed': ('atomId', 'kind'),
    'recommendation_applied': ('atomId', 'slot', 'source'),
    'export_clicked': ('kind', 'atomsCount'),
    'share_created': ('kind', 'score'),
    'feedback_vote': ('atomId', 'vote'),
}
EXPORT_KINDS = ('prompt', 'optimized_prompt', 'recipe', 'storyboard')
SHARE_KINDS = ('check_report', 'recipe_card', 'atom_card', 'storyboard_card')
AID_RE = re.compile(r'^[A-Za-z0-9-]{8,64}$')

_write_lock = threading.Lock()


def filter_event(name, payload):
    fields = EVENT_FIELDS.get(name)
    if fields is None:
        return None
    out = {}
    for k in fields:
        v = payload.get(k) if isinstance(payload, dict) else None
        if v is None:
            continue
        if isinstance(v, str):
            if not v or len(v) > MAX_STR:
                return None
            out[k] = v
        elif isinstance(v, bool):
            continue
        elif isinstance(v, (int, float)):
            if v != v or v in (float('inf'), float('-inf')):  # NaN/inf guard
                return None
            out[k] = v
        elif isinstance(v, list):
            if not v or len(v) > MAX_ARRAY:
                return None
            if any(not isinstance(x, str) or not x or len(x) > MAX_STR for x in v):
                return None
            out[k] = v
    if name == 'export_clicked' and out.get('kind') not in EXPORT_KINDS:
        return None
    if name == 'share_created' and out.get('kind') not in SHARE_KINDS:
        return None
    return out


def append_event(record):
    os.makedirs(DATA_DIR, exist_ok=True)
    day = datetime.now(timezone.utc).strftime('%Y-%m-%d')
    path = os.path.join(DATA_DIR, 'atlas-events-%s.jsonl' % day)
    with _write_lock:
        with open(path, 'a', encoding='utf-8') as f:
            f.write(json.dumps(record, ensure_ascii=False, separators=(',', ':')) + '\n')


class Handler(BaseHTTPRequestHandler):
    protocol_version = 'HTTP/1.1'

    def _cors(self):
        origin = self.headers.get('Origin')
        if origin in ALLOWED_ORIGINS:
            self.send_header('Access-Control-Allow-Origin', origin)
            self.send_header('Vary', 'Origin')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        # deliberately NO Access-Control-Allow-Credentials

    def do_OPTIONS(self):
        self.send_response(204)
        self._cors()
        self.send_header('Content-Length', '0')
        self.end_headers()

    def do_POST(self):
        if urlsplit(self.path).path != '/api/atlas/evt':
            self._reply(404, b'not found')
            return
        length = int(self.headers.get('Content-Length') or 0)
        if length > MAX_BODY:
            self._reply(400, b'too large')
            return
        body = self.rfile.read(length) if length else b''
        try:
            json_body = json.loads(body.decode('utf-8'))
        except Exception:
            self._reply(400, b'bad request')
            return
        if (not isinstance(json_body, dict)
                or not isinstance(json_body.get('aid'), str)
                or not AID_RE.match(json_body['aid'])
                or not isinstance(json_body.get('name'), str)
                or not isinstance(json_body.get('payload'), dict)):
            self._reply(400, b'bad request')
            return
        filtered = filter_event(json_body['name'], json_body['payload'])
        if filtered is None:
            self._reply(422, b'invalid event')
            return
        t = json_body.get('t')
        if not isinstance(t, (int, float)) or isinstance(t, bool):
            t = None
        record = {
            'aid': json_body['aid'],
            't': t if t is not None else int(datetime.now(timezone.utc).timestamp() * 1000),
            'name': json_body['name'],
            'payload': filtered,
            'recvAt': datetime.now(timezone.utc).isoformat(),
        }
        try:
            append_event(record)
        except Exception:
            self._reply(500, b'storage error')
            return
        self.send_response(204)
        self._cors()
        self.send_header('Content-Length', '0')
        self.end_headers()

    def _reply(self, code, msg):
        self.send_response(code)
        self._cors()
        self.send_header('Content-Type', 'text/plain')
        body = msg + b'\n'
        self.send_header('Content-Length', str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, fmt, *args):
        pass  # keep journald quiet per request


if __name__ == '__main__':
    server = ThreadingHTTPServer(('127.0.0.1', PORT), Handler)
    print('atlas-evt-receiver (py) listening on 127.0.0.1:%d (POST /api/atlas/evt)' % PORT, flush=True)
    server.serve_forever()
